lib/ansible/modules/cloud/vmware/vmware_guest_controller.py

```diff
@@ -0,0 +1,553 @@
+                if len(self.disk_ctl_bus_num_list['sata']) != 0:
+                    self.disk_ctl_bus_num_list['sata'].remove(device.busNumber)
+            if isinstance(device, self.nvme_device_type):
+                if len(self.disk_ctl_bus_num_list['nvme']) != 0:
+                    self.disk_ctl_bus_num_list['nvme'].remove(device.busNumber)
+            if isinstance(device, tuple(self.scsi_device_type.values())):
+                if len(self.disk_ctl_bus_num_list['scsi']) != 0:
+                    self.disk_ctl_bus_num_list['scsi'].remove(device.busNumber)
+
+    def check_ctl_disk_exist(self, ctl_type=None, bus_number=None):
+        """
+        Check if controller of specified type exists and if there is disk attaching to it
+        Return: Specified controller device, True or False of attaching disks
+        """
+        ctl_specified = None
+        disks_attached_exist = False
+        if ctl_type is None:
+            return ctl_specified, disks_attached_exist
+
+        for device in self.current_vm_obj.config.hardware.device:
+            if isinstance(device, self.controller_types.get(ctl_type)):
+                if bus_number is not None and device.busNumber != bus_number:
+                    continue
+                ctl_specified = device
+                if len(device.device) != 0:
+                    disks_attached_exist = True
+                break
+
+        return ctl_specified, disks_attached_exist
+
+    def create_controller(self, ctl_type, bus_number=0):
+        """
+        Create new disk or USB controller with specified type
+        Args:
+            ctl_type: controller type
+            bus_number: disk controller bus number
+
+        Return: Virtual device spec for virtual controller
+        """
+        disk_ctl = vim.vm.device.VirtualDeviceSpec()
+        disk_ctl.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
+        if ctl_type == 'sata':
+            disk_ctl.device = self.sata_device_type()
+            disk_ctl.device.key = -randint(15000, 19999)
+        elif ctl_type == 'nvme':
+            disk_ctl.device = self.nvme_device_type()
+            disk_ctl.device.key = -randint(31000, 39999)
+        elif ctl_type in self.scsi_device_type.keys():
+            disk_ctl.device = self.scsi_device_type.get(ctl_type)()
+            disk_ctl.device.key = -randint(1000, 6999)
+            disk_ctl.device.hotAddRemove = True
+            disk_ctl.device.sharedBus = 'noSharing'
+            disk_ctl.device.scsiCtlrUnitNumber = 7
+        elif ctl_type in self.usb_device_type.keys():
+            disk_ctl.device = self.usb_device_type.get(ctl_type)()
+            if ctl_type == 'usb2':
+                disk_ctl.device.key = 7000
+            elif ctl_type == 'usb3':
+                disk_ctl.device.key = 14000
+
+        disk_ctl.device.deviceInfo = vim.Description()
+        disk_ctl.device.busNumber = bus_number
+
+        return disk_ctl
+
+    def gather_disk_controller_facts(self):
```

Comment:
As a convention in Ansible modules, this can be separate _info module? @Akasurde , Your thoughts?

Reply:
Yes, we can write a info module in separate PR.

Reply:
Yeah, if there are requests, we can add a new module "vmware_guest_controller_info", but in my opinion, this can also be implemented in one module. Thanks.

Reply:
OK. We can add info module as separate PR